File: BACKEND/SCRIPT/SEQ_ONTO/wh_seq_ontol.php
<?php


/**
 SCRIPT NAME: wh_seq_ontol
 PURPOSE:     Process the sequence ontology file and update the database
 
*/
$JOB_NAME='wh_seq_ontol';

/// Get root directories
$TG_DIR= getenv('TG_DIR');
if ($TG_DIR===false)  die('NO TG_DIR found ');
if (!is_dir($TG_DIR)) die('TG_DIR value is not a directory '.$TG_DIR);
require_once($TG_DIR.'/BACKEND/SCRIPT/LIB/loader.php');

/// Get job id
$JOB_ID=getJobIDByName($JOB_NAME);
$PROCESS_CONTROL['DIR']='N/A';
/// Get job info
$JOB_INFO=$GLB_TREE[$JOB_ID];




addLog("Check directory");

	/// Get parent job information
	$CK_SEQ_ONTOL_INFO=$GLB_TREE[getJobIDByName('ck_seq_ontol')];
	
	/// Set up directory
	$W_DIR=$TG_DIR.'/'.$GLB_VAR['PROCESS_DIR'].'/'.$JOB_INFO['DIR'].'/'.$CK_SEQ_ONTOL_INFO['TIME']['DEV_DIR'];	
	if (!is_dir($W_DIR)) 															failProcess($JOB_ID."002",'NO '.$W_DIR.' found ');
	if (!chdir($W_DIR))																failProcess($JOB_ID."003",'Unable to chdir to '.$W_DIR);

	/// Update process control directory to the current release so that the next job can use it
	$PROCESS_CONTROL['DIR']=$CK_SEQ_ONTOL_INFO['TIME']['DEV_DIR'];

	$F_FILE=$W_DIR.'/so.obo';				if (!checkFileExist($F_FILE))			failProcess($JOB_ID."004",'NO '.$F_FILE.' found ');
	
	$STATS=array('SO'=>0);



addLog("Loading existing data from table");
	

	$SQL='SELECT so_entry_Id, so_id, so_name,so_description FROM so_entry';
	$res=runQuery($SQL);
	if ($res===false)															failProcess($JOB_ID."005",'Return code failure ');

	/// Push all records from the database into $SEQ_ONTOL array
	$SEQ_ONTOL=array();
	$MAX_DBID=0;
	foreach ($res as $tab)
	{
		
		$SEQ_ONTOL[$tab['so_id']]=array(
			'DB'=>$tab['so_entry_id'],
			'NAME'=>$tab['so_name'],
			'DESC'=>$tab['so_description'],
			'STATUS'=>'DB');

			/// Get the max id
		if ($tab['so_entry_id']>$MAX_DBID)$MAX_DBID=$tab['so_entry_id'];
	}
	unset($res);
	
	//print_r($SEQ_ONTOL);

addLog("Process file");

	$fp=fopen($F_FILE,'r');	if (!$fp)												failProcess($JOB_ID."006",'Unable to open '.$F_FILE);
	while(!feof($fp))
	{
		$line=stream_get_line($fp,10000,"\n");
		/// A record in the file is a block of line, starting with [Term] and end with an empty line
		if ($line!='[Term]')continue;
		$STATS['SO']++;
		$CURR_ID=-1;
		$so_id='';
		do
		{
			$line=stream_get_line($fp,10000,"\n");
			if ($line=='')break;/// Emtpy line -> end of record
			
			/// Extract header from value
			$pos=strpos($line,':');
			
			$head=substr($line,0,$pos);
			$val=substr($line,$pos+2);

			/// Lookup the header to check if the record exist in the dabtabase
			if ($head=='id')
			{
				$so_id=$val;
				$CURR_ID=$so_id;
				/// Not found -> Create it
				if (!isset($SEQ_ONTOL[$so_id]))
				{
					++$MAX_DBID;
					$SEQ_ONTOL[$CURR_ID]=array('DB'=>$MAX_DBID,'NAME'=>'','DESC'=>'','STATUS'=>'NEW');
				}
				/// Already exist -> set to valid
				else $SEQ_ONTOL[$CURR_ID]['STATUS']='VALID';
			}
		
			else if ($head=='name')
			{
				/// Name is different -> set to UPD, unless it's a new record
				if ($SEQ_ONTOL[$CURR_ID]['NAME']!=$val)
				{
					$SEQ_ONTOL[$CURR_ID]['NAME']=$val;
					if ($SEQ_ONTOL[$CURR_ID]['STATUS']!='NEW')
					{
						$SEQ_ONTOL[$CURR_ID]['STATUS']='UPD';
					}
				}
			}
			else if ($head=='def')
			{
				/// definition is different -> set to UPD, unless it's a new record
				$pos=strrpos($val,'"');
				$val=substr($val,1,$pos-1);
				if ($SEQ_ONTOL[$CURR_ID]['STATUS']=='NEW')
				{
					$SEQ_ONTOL[$CURR_ID]['DESC']=$val;
				}
				else if ($SEQ_ONTOL[$CURR_ID]['DESC']!=$val)
				{
					$SEQ_ONTOL[$CURR_ID]['DESC']=$val;
					$SEQ_ONTOL[$CURR_ID]['STATUS']='UPD';
				}
			}
			
			
		}while(!feof($fp));
		$so_id='';

	}
	fclose($fp);
	


addLog("Update records");
	$fp=fopen('so_insert.csv','w');	if(!$fp)								failProcess($JOB_ID."007",'Unable to open so_insert file');
	

	/// Now we have to update the records that need to be updated and push the new records into a file
	foreach ($SEQ_ONTOL as $so_id=>$INFO)
	{

		// Update the records that need to be updated
			if ($INFO['STATUS']=='UPD')
			{
				$QUERY="UPDATE so_entry 
				SET so_name='".str_replace("'","''",$INFO['NAME'])."', 
				so_description='".str_replace("'","''",$INFO['DESC'])."',
				so_id='".$so_id."'
				WHERE so_entry_id =".$INFO['DB'];
				if (!runQueryNoRes($QUERY))								failProcess($JOB_ID."008",'Unable to run query '.$QUERY);
			}
		/// Push into a file all new recods
			else if ($INFO['STATUS']=='NEW')
			{
				fputs($fp,
					$INFO['DB']."\t".
					$INFO['NAME']."\t".
					$so_id."\t".
					(($INFO['DESC']=='')?'NULL':'"'.str_replace('"','""',$INFO['DESC']).'"')."\n");
			}
	}

	fclose($fp);


addLog("Delete records");
/// Delete those that haven't been found in the file
	foreach ($SEQ_ONTOL as $ID=>$INFO)
	{
		if ($INFO['STATUS']!='DB')continue;///DELETION
	
		$query='DELETE FROM SO_ENTRY WHERE SO_ENTRY_ID='.$INFO['DB'].'"';
		if (!runQueryNoRes($query))										failProcess($JOB_ID."009",'Unable to run query '.$query);
		
	}

	/// Push the new records in the database
	$FCAV_NAME='so_insert.csv';
	$command='\COPY '.$GLB_VAR['DB_SCHEMA'].'.so_entry(so_entry_id,so_name,so_id,so_description)FROM \''.$FCAV_NAME."'  (DELIMITER E'\\t', null \\\"NULL\\\" ,format CSV )";
	echo $DB_INFO['COMMAND'].' -c "'.$command.'"'."\n";
	system($DB_INFO['COMMAND'].' -c "'.$command.'"',$return_code);
	if ($return_code !=0 )												failProcess($JOB_ID."010",'Unable to insert so_entry'); 


	$CLEANUP=array('so_insert.csv');
	foreach ($CLEANUP as $file) 
	{
		if (!is_file($file))continue;
		if (!unlink($file))											failProcess($JOB_ID."011",'Unable to delete '.$file);
	}


	updateStat('so_entry','so_entry',$STATS['SO'],$JOB_ID);


addLog("Push to prod");
	pushToProd();

	

successProcess();

?>

File: BACKEND/SCRIPT/SEQ_ONTO/wh_seq_ontol.py
import os


# Get root directories
TG_DIR = os.getenv('TG_DIR')
if TG_DIR is None:
    print('NO TG_DIR found')
    exit(1)
if not os.path.isdir(TG_DIR):
    print('TG_DIR value is not a directory ')
    exit(2)



import sys
sys.path.append(os.getenv('TG_DIR')+"/BACKEND/SCRIPT/LIB_PYTHON/")
from fct_utils import *
from loader import *
import csv
import pprint

job_name = 'wh_seq_ontol'
JOB_ID = get_job_id_by_name(job_name,True)
JOB_INFO=GLB_TREE[JOB_ID]
PROCESS_CONTROL['JOB_NAME'] =job_name

pprint.pp(JOB_INFO)
add_log("Check directory")

# Get parent job information
parent_job_name = 'ck_seq_ontol'
PARENT_JOB_ID = get_job_id_by_name(parent_job_name,True)
CK_SEQ_ONTOL_INFO = GLB_TREE[PARENT_JOB_ID]



# Set up directory
W_DIR = os.path.join(TG_DIR, GLB_VAR['PROCESS_DIR'], JOB_INFO['DIR'], CK_SEQ_ONTOL_INFO['TIME']['DEV_DIR'])
if not os.path.isdir(W_DIR):
    fail_process(JOB_ID + "002", 'NO ' + W_DIR + ' found ')
os.chdir(W_DIR)
if (os.getcwd() != W_DIR):
    fail_process(JOB_ID + "003", 'Unable to chdir to ' + W_DIR)

# Update process control directory
PROCESS_CONTROL['DIR'] = CK_SEQ_ONTOL_INFO['TIME']['DEV_DIR']

F_FILE = os.path.join(W_DIR, 'so.obo')
if not os.path.isfile(F_FILE):
    fail_process(JOB_ID + "004", 'NO ' + F_FILE + ' found ')

STATS = {'SO': 0}

add_log("Loading existing data from table")

# Assuming runQuery is defined somewhere
SQL = 'SELECT so_entry_Id, so_id, so_name, so_description FROM so_entry'
res = run_query(SQL)
if res is False:
    fail_process(JOB_ID + "005", 'Return code failure ')

SEQ_ONTOL = {}
MAX_DBID = 0
for tab in res:
    SEQ_ONTOL[tab['so_id']] = {
        'DB': tab['so_entry_id'],
        'NAME': tab['so_name'],
        'DESC': tab['so_description'],
        'STATUS': 'DB'
    }
    if tab['so_entry_id'] > MAX_DBID:
        MAX_DBID = tab['so_entry_id']
del res

print("Number of records from db: "+str(len(SEQ_ONTOL)))

add_log("Process file")

with open(F_FILE, 'r') as fp:
    for line in fp:
        if line.strip() != '[Term]':
            continue
        
        STATS['SO'] += 1
        CURR_ID = -1
        so_id = ''
        for line in fp:
            if line.strip() == '':
                break
            pos = line.find(':')
            head = line[:pos]
            val = line[pos+2:].strip()
            if head == 'id':
                so_id = val
                CURR_ID = so_id
                if so_id not in SEQ_ONTOL:
                    MAX_DBID += 1
                    
                    SEQ_ONTOL[CURR_ID] = {'DB': MAX_DBID, 'NAME': '', 'DESC': '', 'STATUS': 'NEW'}
                else:
                    SEQ_ONTOL[CURR_ID]['STATUS'] = 'VALID'
            elif head == 'name':
                if SEQ_ONTOL[CURR_ID]['NAME'] != val:
                    SEQ_ONTOL[CURR_ID]['NAME'] = val
                    if SEQ_ONTOL[CURR_ID]['STATUS'] != 'NEW':
                        SEQ_ONTOL[CURR_ID]['STATUS'] = 'UPD'
            elif head == 'def':
                pos = val.rfind('"')
                val = val[1:pos]
                if SEQ_ONTOL[CURR_ID]['STATUS'] == 'NEW':
                    SEQ_ONTOL[CURR_ID]['DESC'] = val
                elif SEQ_ONTOL[CURR_ID]['DESC'] != val:
                    SEQ_ONTOL[CURR_ID]['DESC'] = val
                    SEQ_ONTOL[CURR_ID]['STATUS'] = 'UPD'

add_log("Update records")


with open(os.getcwd()+'/so_insert.csv', 'w') as fp:
    writer = csv.writer(fp, delimiter='\t')
    for so_id, info in SEQ_ONTOL.items():
        if info['STATUS'] == 'UPD':
            query = f"UPDATE so_entry SET so_name='{info['NAME'].replace("'", "''")}', " \
                    f"so_description='{info['DESC'].replace("'", "''")}', so_id='{so_id}' " \
                    f"WHERE so_entry_id={info['DB']}"
            print(query)
            if not run_query_no_res(query):
                fail_process(JOB_ID + "008", 'Unable to run query ' + query)
        elif info['STATUS'] == 'NEW':
            pprint.pp(info)
            writer.writerow([info['DB'], info['NAME'], so_id, info['DESC'] if info['DESC'] else 'NULL'])

add_log("Delete records")

for ID, info in SEQ_ONTOL.items():
    if info['STATUS'] != 'DB':
        continue
    query = f"DELETE FROM SO_ENTRY WHERE SO_ENTRY_ID={info['DB']}"
    print(query)
    if not run_query_no_res(query):
        fail_process(JOB_ID + "009", 'Unable to run query ' + query)

FCAV_NAME = 'so_insert.csv'
command = f"\\COPY {GLB_VAR['DB_SCHEMA']}.so_entry(so_entry_id,so_name,so_id,so_description) " \
          f"FROM '{FCAV_NAME}' (DELIMITER E'\\t', null 'NULL', format CSV )"
return_code=os.system(f"{DB_INFO['COMMAND']} -c \"{command}\"")
if return_code != 0:
    fail_process(JOB_ID + "010", 'Unable to insert sequence_ontology entry')





CLEANUP = ['so_insert.csv']
for file in CLEANUP:
    if os.path.isfile(file):
        os.unlink(file)

update_stat('so_entry', 'so_entry', STATS['SO'], JOB_ID)
push_to_prod(JOB_INFO,W_DIR)

success_process()
